src/flowfield.cpp

```diff
@@ -0,0 +1,811 @@
+	return true;
+}
+
+
+void ffpathShutdown()
+{
+	if (ffpathThread)
+	{
+		// Signal the path finding thread to quit
+		ffpathQuit = true;
+		wzSemaphorePost(ffpathSemaphore);  // Wake up thread.
+
+		wzThreadJoin(ffpathThread);
+		ffpathThread = nullptr;
+		wzMutexDestroy(ffpathMutex);
+		ffpathMutex = nullptr;
+		wzSemaphoreDestroy(ffpathSemaphore);
+		ffpathSemaphore = nullptr;
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+inline unsigned int coordinateToArrayIndex(unsigned short x, unsigned short y) { return y * FF_MAP_WIDTH + x; }
+inline Vector2i arrayIndexToCoordinate(unsigned int index) { return Vector2i { index % FF_MAP_WIDTH, index / FF_MAP_WIDTH }; }
+
+struct IntegrationField {
+	unsigned short cost[FF_MAP_AREA];
+	void setCost(unsigned int x, unsigned int y, unsigned short value){
+		this->cost[coordinateToArrayIndex(x, y)] = value;
+	}
+	unsigned short getCost(unsigned int x, unsigned int y){
+		return this->cost[coordinateToArrayIndex(x, y)];
+	}
+	void setCost(unsigned int index, unsigned short value){
+		this->cost[index] = value;
+	}
+	unsigned short getCost(unsigned int index){
+		return this->cost[index];
+	}
+};
+
+struct CostField {
+	unsigned short cost[FF_MAP_AREA];
+	void setCost(unsigned int x, unsigned int y, unsigned short value){
+		this->cost[coordinateToArrayIndex(x, y)] = value;
+	}
+	unsigned short getCost(unsigned int x, unsigned int y){
+		return this->cost[coordinateToArrayIndex(x, y)];
+	}
+	void setCost(unsigned int index, unsigned short value){
+		this->cost[index] = value;
+	}
+	unsigned short getCost(unsigned int index){
+		return this->cost[index];
+	}
+};
+
+// TODO: Remove this in favor of glm.
+struct VectorT {
+	float x;
+	float y;
+
+	void normalize() {
+		const float length = std::sqrt(std::pow(x, 2) + std::pow(y, 2));
+
+		if (length != 0) {
+			x /= length;
+			y /= length;
+		}
+	}
+};
+
+/// auto increment state for flowfield ids. Used on psDroid->sMove.flowfield. 0 means no flowfield exists for the unit.
+unsigned int flowfieldIdIncrementor = 1;
+
+struct Flowfield {
+	unsigned int id;
+	std::vector<Vector2i> goals;
+	std::unique_ptr<IntegrationField> integrationField;
+	std::array<VectorT, FF_MAP_AREA> vectors;
+
+	void setVector(unsigned short x, unsigned short y, VectorT vector) {
+		vectors[coordinateToArrayIndex(x, y)] = vector;
+	}
+	VectorT getVector(unsigned short x, unsigned short y) const {
+		return vectors[coordinateToArrayIndex(x, y)];
+	}
+};
+
+// Cost fields for ground, hover and lift movement types
+std::array<std::unique_ptr<CostField>, 4> costFields {
+	std::unique_ptr<CostField>(new CostField()),
+	std::unique_ptr<CostField>(new CostField()),
+	std::unique_ptr<CostField>(new CostField()),
+	std::unique_ptr<CostField>(new CostField()),
+};
+
+// Flow field cache for ground, hover and lift movement types
+std::array<std::unique_ptr<std::map<std::vector<ComparableVector2i>, std::unique_ptr<Flowfield>>>, 4> flowfieldCaches {
```

Comment:
Is there a limit for the cache size or does it eat up memory until each possible pair of start-end is in the cache?

Reply:
Currently in eat buffet mode 😆